hail/python/hailtop/aiotools/fs.py

```diff
@@ -0,0 +1,164 @@
+from typing import TypeVar, Optional, List, Type, BinaryIO, cast, Set
+from types import TracebackType
+import abc
+import os
+import os.path
+import shutil
+from concurrent.futures import ThreadPoolExecutor
+import urllib.parse
+from hailtop.utils import blocking_to_async
+from .stream import AsyncStream, blocking_stream_to_async
+
+AsyncFSType = TypeVar('AsyncFSType', bound='AsyncFS')
+
+
+class AsyncFS(abc.ABC):
+    @abc.abstractmethod
+    def schemes(self) -> Set[str]:
+        pass
+
+    @abc.abstractmethod
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        pass
+
+    @abc.abstractmethod
+    async def mkdir(self, url: str) -> None:
+        pass
+
+    @abc.abstractmethod
+    async def isfile(self, url: str) -> bool:
+        pass
+
+    @abc.abstractmethod
+    async def isdir(self, url: str) -> bool:
+        pass
+
+    @abc.abstractmethod
+    async def remove(self, url: str) -> None:
+        pass
+
+    @abc.abstractmethod
+    async def rmtree(self, url: str) -> None:
+        pass
+
+    async def touch(self, url: str) -> None:
+        async with await self.open(url, 'wb'):
+            pass
+
+    async def close(self) -> None:
+        pass
+
+    async def __aenter__(self: AsyncFSType) -> AsyncFSType:
+        return self
+
+    async def __aexit__(self,
+                        exc_type: Optional[Type[BaseException]],
+                        exc_val: Optional[BaseException],
+                        exc_tb: Optional[TracebackType]) -> None:
+        await self.close()
+
+
+class LocalAsyncFS(AsyncFS):
+    def __init__(self, thread_pool: ThreadPoolExecutor, max_workers=None):
+        if not thread_pool:
+            thread_pool = ThreadPoolExecutor(max_workers=max_workers)
+        self._thread_pool = thread_pool
+
+    def schemes(self) -> Set[str]:
+        return {'file'}
+
+    @staticmethod
+    def _get_path(url):
+        parsed = urllib.parse.urlparse(url)
+        if parsed.scheme and parsed.scheme != 'file':
+            raise ValueError(f"invalid scheme, expected file: {parsed.scheme}")
+        return parsed.path
+
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        if 'b' not in mode:
```

Comment:
I agree.  That's part of why I said "I'm skeptical about pushing hard on [copying the standard Python interface]".  Let me change this now.  I'm going to have:

```
  def open(self, url: str) -> AsyncStream:
    ...

  def create(self, url: str) -> AsyncStream
    ...
```

And the `mode` argument will go away.  If I do this, I can also split AsyncStream into `ReadableStream` and `WritableStream`.  That's better than the current design, because we won't support read-and-writable streams.